### DISCUSSION TITLE
Parsing anchored, earliest match from `BufRead`

### DISCUSSION THREAD

**Discussion by s5bug** [rust-lang/regex#1186] [Q&A]
I'm quite new to Rust, so please let me know if there's something I'm missing (a lot of this is pseudocode and doesn't actually compile, but I hope it illustrates my point):

I have a dead-simple pattern and I want to grab it from a `BufRead`, and here's what I have so far:
```rs
const READ_REGEX: dense::DFA<Vec<u32>> = dense::DFA::builder()
    .configure(dense::DFA::config().start_kind(StartKind::Anchored))
    .build(r"\u001bP(\d+)!~([0-9a-fA-F]{4})\u001b\\")?;

pub fn read_from<A>(inp: &mut A) -> DecCkSR where A: BufRead {
    let buf = inp.fill_buf().unwrap();
    let input = Input::new(buf)
        .anchored(Anchored::Yes)
        .earliest(true);
```

So what I need to do is:
- when I've received input that means the DFA will never match, I need to error out
- when I've received a match, I want to grab the contents of the capture groups, and then `inp.consume` by the length of the match
- when I've received input that could match in the future, but isn't matching right now, I want to `inp.consume` it and then continue advancing with another `inp.fill_buf`

Is there a better way to do this than advancing the DFA on bytes?

**Comment by BurntSushi** [comment]
You probably have to match byte at a time. But for capture groups, you probably need the full match to be in a contiguous region of memory. Otherwise, it seems like you want streaming searches, which is a big can of worms. Search the issue tracker. (I'm on mobile or else I would link the relevant issues.)

Also, your code doesn't look like it will compile...

**Reply by s5bug** [reply]
Yeah, it doesn't compile, still working on it....

I looked at the issues first, which is why I decided I needed `regex_automaton`. I don't mind throwing memory into a growable buffer; this is one of those cases where the only input this application will be receiving is terminal keystrokes so if it runs out of memory that's 100% the user's fault.
